source/language/types.rst

```diff
@@ -368,54 +510,112 @@ their argument expression in parentheses:
       +----------+-----------------------------------+--------------------------------------+----------------------------------------+
-      | sqrt     | ``float``                         | ``float``                            | Returns the principal root.            |
-      |          |                                   |                                      |                                        |
+      | sqrt     | ``float``                         | ``float``                            | Square root.  This always returns the  |
+      |          |                                   |                                      | principal root.                        |
       |          | ``complex``                       | ``complex``                          |                                        |
       +----------+-----------------------------------+--------------------------------------+----------------------------------------+
-      | tan      | (``float`` or ``angle``)          | ``float``                            |                                        |
+      | tan      | (``float`` or ``angle``)          | ``float``                            | Tangent.                               |
       +----------+-----------------------------------+--------------------------------------+----------------------------------------+
 
+For each built-in function, the chosen overload is the first one to appear in
+the list above where all given operands can be implicitly cast to the valid
+input types.  The output type is not considered when choosing an overload.  It
+is an error if there is no valid overload for a given sequence of operands.
+
+.. code-block::
+
+   // Valid statements.
+
+   const float[64] f1 = 2.5;
+   const int[8] i1 = 4;
+   const uint[4] u1 = 3;
+   const bit[8] b1 = "0010_1010";
+   const complex[float[64]] c1 = 1.0 + 2.0im;
+
+   const float[64] f2 = 2.0 * exp(f1);
+   const float[64] f3 = exp(i1);
+   // The ``float -> float`` overload of ``exp`` is chosen in both of these
+   // cases; in the first, there is an exact type match, in the second the
+   // ``int[8]`` input can be implicitly promoted to ``float``.
+
+   const int[8] i2 = pow(i1, u1);
+   // Value 64, expression has type `const int`.  The first overload of `pow`
+   // is chosen, because `i1` can be implicitly promoted to `const int` and
+   // `u1` to `const uint`.
+
+   const float[64] f4 = pow(i1, -2);
+   // Value 0.0625, expression has type `const float`.  The second,
+   // `(float, float) -> float`, overload is chosen, because `-2` (type
+   // `const int`) cannot be implicitly promoted to `const uint`, but both
+   // input types can be implicitly promoted to `float`.  The `complex` overload
+   // is not attempted, because it has lower priority.
+
+   const bit[8] b2 = rotl(b1, 3);
+   // Value "0101_0001", expression has type `const bit[8]` (the 
```

Comment:
Maybe also add an example for what happens with a negative integer?

Reply:
No longer possible after https://github.com/openqasm/openqasm/pull/366/commits/307fa094e8be792a845623101323999a452ab261